<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title>Protein Annotation of PubMed XML Abstracts with GeneTUKit</title>
  <style type="text/css">code{white-space: pre;}</style>
</head>
<body>
<div id="header">
<h1 class="title">Protein Annotation of PubMed XML Abstracts with GeneTUKit</h1>
</div>
<div id="TOC">
<ul>
<li><a href="#introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#download-of-pubmed-abstracts"><span class="toc-section-number">2</span> Download of PubMed Abstracts</a></li>
<li><a href="#geneprotein-extraction-wit-genetukit"><span class="toc-section-number">3</span> Gene/Protein Extraction wit GeneTUKit</a></li>
<li><a href="#modifying-the-genetukit-output"><span class="toc-section-number">4</span> Modifying the GeneTUKit output</a></li>
<li><a href="#mapping-of-uniprot-ids-to-gene-ids"><span class="toc-section-number">5</span> Mapping of UniProt IDs to Gene IDs</a></li>
<li><a href="#annotation-of-pubmed-abstracts"><span class="toc-section-number">6</span> Annotation of PubMed Abstracts</a></li>
<li><a href="#contact"><span class="toc-section-number">7</span> Contact</a></li>
</ul>
</div>
<h1 id="introduction"><a href="#introduction"><span class="header-section-number">1</span> Introduction</a></h1>
<ul>
<li>The following pipeline annotates genes/proteins in a set of PubMed abstracts with GeneTUKit.</li>
<li>It uses PubMed2Go to build a PostgreSQL database from PubMed XML files that can be downloaded from NCBI.</li>
<li><p>PubMed2Go was renamed to PubMedPortable. It is available via the old and the new URLs:</p>
<blockquote>
<ul>
<li><a href="https://github.com/KerstenDoering/PubMed2Go">https://github.com/KerstenDoering/PubMed2Go</a></li>
<li><a href="https://github.com/KerstenDoering/PubMedPortable">https://github.com/KerstenDoering/PubMedPortable</a></li>
</ul>
</blockquote></li>
<li>Start by copying the whole project folder from GitHub to your local disk.</li>
<li>The pipeline was tested on a Linux system with Ubuntu 14.04.</li>
<li>The following flow chart shows the order of the pipeline scripts and their input and output files.</li>
</ul>
<p><img src="flowchart.jpg" alt="image" /></p>
<h1 id="download-of-pubmed-abstracts"><a href="#download-of-pubmed-abstracts"><span class="header-section-number">2</span> Download of PubMed Abstracts</a></h1>
<ul>
<li><p>The title and abstract part of requested PubMed XML documents can be accessed with PubMedPortable:</p>
<blockquote>
<ul>
<li><a href="https://github.com/KerstenDoering/PubMedPortable/wiki">https://github.com/KerstenDoering/PubMedPortable/wiki</a></li>
</ul>
</blockquote></li>
<li><p>The PubMedPortable documentation shows how to download a data set and build a PostgreSQL relational database. Follow the installation instructions. The part describing how to build a Xapian full text index is not needed here:</p>
<blockquote>
<ul>
<li><a href="https://github.com/KerstenDoering/PubMedPortable/blob/master/documentation/quick_install.rst">https://github.com/KerstenDoering/PubMedPortable/blob/master/documentation/quick_install.rst</a></li>
</ul>
</blockquote></li>
<li>This documentation refers to the PubMedPortable example data set processing texts which deal with the disease pancreatic cancer.</li>
<li><p>The script download_articles.py reads the PubMed IDs in the file pubmed_result.txt. Four randomly selected abstracts with the PubMed IDs 10025831, 23215050, 24622518, and 24842107 were selected from the PostgreSQL database and saved in the folder downloaded_abstracts in pseudo XML format using this script.</p>
<blockquote>
<ul>
<li>The script can be run with “python download_articles.py” and modified with the name of the database, output folder, and input file (use “python download_articles.py -h” for more information).</li>
</ul>
</blockquote></li>
<li>Unfortunately, the plain text parameter in GeneTUKit does not work. That is the reason for using pseudo XML files. These XML tags are required to run GeneTUKit in XML mode and to recognise the title and text separately.</li>
<li>The following steps cannot only be applied to PubMed XML abstracts, but to any texts that are given in NXML format.</li>
<li><p>NXML is a data structure used in BioCreAtIvE III and the example on the GeneTUKit homepage refers to this document:</p>
<blockquote>
<ul>
<li><a href="http://www.qanswers.net/download/genetukit/1934391.nxml">http://www.qanswers.net/download/genetukit/1934391.nxml</a></li>
</ul>
</blockquote></li>
<li>The mandatory XML structure used by GeneTUKit is &lt;article&gt;&lt;article-meta&gt;&lt;article-title&gt; TITLE &lt;/article-title&gt;&lt;abstract&gt;&lt;p&gt; ABSTRACT &lt;/p&gt;&lt;/abstract&gt;&lt;/article-meta&gt;&lt;/article&gt;.</li>
<li>If no abstract text is provided by PubMed, the structure is &lt;article&gt;&lt;article-meta&gt;&lt;abstract&gt;&lt;p&gt; TITLE &lt;/p&gt;&lt;/abstract&gt;&lt;/article-meta&gt;&lt;/article&gt;. In this special case, the title has to be encoded within the abstract tags to be recognised by GeneTUKit.</li>
<li>In XML documents, certain characters are not allowed in normal text, e.g. “&lt;” and “&amp;”. Therefore they are escaped by the script and replaced with their ASCII codes.</li>
</ul>
<h1 id="geneprotein-extraction-wit-genetukit"><a href="#geneprotein-extraction-wit-genetukit"><span class="header-section-number">3</span> Gene/Protein Extraction wit GeneTUKit</a></h1>
<ul>
<li>The extraction of gene/protein names is done by running GeneTUKit and the output is saved in a CSV file.</li>
<li>The latest version of GeneTUKit can be freely downloaded from <a href="http://www.qanswers.net/GeneTUKit/download.html">http://www.qanswers.net/GeneTUKit/download.html</a>.</li>
<li>There is also a link providing the latest EntrezGene data set (gene_info.gz). This file has to be downloaded and copied into the EntrezGene directory of the downloaded GeneTUKit folder.</li>
<li>To install GeneTUKit, step one and three from the official help page (<a href="http://www.qanswers.net/GeneTUKit/help.html">http://www.qanswers.net/GeneTUKit/help.html</a>) are mandatory for using the pipeline presented herein.</li>
<li><p>In this pipeline, the GeneTUKit installation with the default MySQL version was tested. Therefore, the GeneTUKit config file does not have to be changed.</p>
<blockquote>
<ul>
<li>It is possible to connect to a MySQL database via command-line with “mysql -u root -p”.</li>
</ul>
</blockquote></li>
<li>Install the latest MySQL database client and server version. Although the GeneTUKit documentation recommends the extra download of the jdbc driver JAR package for this database engine, this was not needed in the test system with Ubuntu 14.04 LTS.</li>
<li>Copy the folder downloaded_abstracts with the NXML files generated in the last section into the new GeneTUKit directory.</li>
<li><p>GeneTUKit is using CRF++ for NER (Name Entity Recognition), but the link provided by the just mentioned GeneTUKit page leads to a repository requiring a password. It’s also possible to use BANNER for this purpose. Running GeneTUKit with a single NXML file (with BANNER) works as follows:</p>
<blockquote>
<ul>
<li>java -Djava.library.path=. -jar genetukit.jar -x downloaded_abstracts/23215050.nxml -banner</li>
</ul>
</blockquote></li>
<li><p>If you want to process multiple files at ones, use the batch mode by providing the folder name instead of a file name:</p>
<blockquote>
<ul>
<li>java -Djava.library.path=. -jar genetukit.jar -x downloaded_abstracts -banner</li>
</ul>
</blockquote></li>
<li><p>With “&gt;”, the output is piped to the output file gtk_output.csv and the Java library path can probably be omitted:</p>
<blockquote>
<ul>
<li>java -jar genetukit.jar -x downloaded_abstracts -banner &gt; gtk_output.csv</li>
</ul>
</blockquote></li>
<li>The running time for this step on a test data set of 5000 abstracts was around 5 h with a 2 GHz single core CPU.</li>
</ul>
<h1 id="modifying-the-genetukit-output"><a href="#modifying-the-genetukit-output"><span class="header-section-number">4</span> Modifying the GeneTUKit output</a></h1>
<ul>
<li><p>From the output given by GeneTUKit (e.g. GeneTUKit/gtk_output.csv), the required values PubMed ID, gene ID, and the synonyms are saved in a new CSV file named pmid_geneid_syn.csv (“python filter_out_genetukit_output.py”, executed from the original GitHub project folder GeneTUKit-Pipeline)</p>
<blockquote>
<ul>
<li>The default parameters can be shown with parameter “-h”.</li>
</ul>
</blockquote></li>
<li>GeneTUKit also provides an organism ID and a score for each prediction’s certainty. These values are not further processed in this script, but it is reasonable to consider especially the prediction score.</li>
</ul>
<h1 id="mapping-of-uniprot-ids-to-gene-ids"><a href="#mapping-of-uniprot-ids-to-gene-ids"><span class="header-section-number">5</span> Mapping of UniProt IDs to Gene IDs</a></h1>
<ul>
<li><p>Each gene ID provided by GeneTUKit has to be mapped to its respective UniProt ID. Using UniProt IDs brings up the advantage of being able to access their sequenes (<a href="http://www.uniprot.org">http://www.uniprot.org</a>). The UniProt IDs are contained in idmapping.dat.gz. This file has to be extracted in the main project folder (GeneTUKit-Pipeline) and can be downloaded here:</p>
<blockquote>
<ul>
<li><a href="ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/idmapping/">ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/idmapping/</a></li>
<li>The file idmapping.dat.gz also contains some unrelated information. By running filter_idmapping.py with the option “-t”, only the related gene IDs and UniProt IDs from pmid_geneid_syn.csv are saved in “filtered_idmapping.csv”. The converted file is also saved in the main directory.</li>
<li>The option of using the test case with a small number of gene IDs was used to complete the example presented in this documentation. In general, all gene IDs are needed to extract the appropriate UniProt ID from idmapping.dat.</li>
</ul>
</blockquote></li>
<li>The mapping process of storing PubMed ID, (mapped) gene ID, synonym(s), and UniProt ID(s) for each different synonym is executed with map_geneid_to_uniprotid.py.</li>
<li><p>The script reads the file filtered_idmapping.csv by importing the script map_to_dict.py, which creates a dictionary data structure, containing the gene IDs as keys and the UniProt IDs as values.</p>
<blockquote>
<ul>
<li>This dictionary is used to create the output file merged_file.csv with one line per synonym (tab-separated): PubMed ID, gene ID, synonym, UniProt ID(s).</li>
<li>The script also creates a second file is which is a dictionary of dictionaries from PubMed IDs, containing all synonyms of an abstract as keys with all UniProt IDs as values. This output is stored as the Python pickle file “save.p” to be used within the next pipeline step.</li>
</ul>
</blockquote></li>
<li><p>The script can be run without additional parameters:</p>
<blockquote>
<ul>
<li>python map_geneid_to_uniprotid.py</li>
</ul>
</blockquote></li>
</ul>
<h1 id="annotation-of-pubmed-abstracts"><a href="#annotation-of-pubmed-abstracts"><span class="header-section-number">6</span> Annotation of PubMed Abstracts</a></h1>
<ul>
<li><p>The gene/protein tags are added to the provided XML files with the following command:</p>
<blockquote>
<ul>
<li>python annotate_abstracts.py -i downloaded_abstracts</li>
</ul>
</blockquote></li>
<li>The script takes the path to the downloaded pseudo XML texts specified by the parameter “-i” and the list of synonym-UniProt ID pairs saved in the dictionary save.p from the last step. The tagged abstract titles and texts are saved tab-separated in a CSV file named annotated_abstracts.csv, each row a new PubMed ID (without pseudo XML tags).</li>
<li>All abstract texts and titlse are separately searched for each synonym. The implementation takes care for nested tags in a way that it only highlights the longest matching synonym (function remove_nested_tagging()).</li>
</ul>
<h1 id="contact"><a href="#contact"><span class="header-section-number">7</span> Contact</a></h1>
<ul>
<li><p>Please, write an e-mail, if you have questions, feedback, improvements, or new ideas:</p>
<blockquote>
<ul>
<li><script type="text/javascript">
<!--
h='&#x79;&#x61;&#104;&#x6f;&#x6f;&#46;&#x63;&#x6f;&#x6d;';a='&#64;';n='&#x65;&#x5f;&#x61;&#98;&#98;&#x61;&#x73;&#x69;&#x61;&#110;';e=n+a+h;
document.write('<a h'+'ref'+'="ma'+'ilto'+':'+e+'">'+e+'<\/'+'a'+'>');
// -->
</script><noscript>&#x65;&#x5f;&#x61;&#98;&#98;&#x61;&#x73;&#x69;&#x61;&#110;&#32;&#x61;&#116;&#32;&#x79;&#x61;&#104;&#x6f;&#x6f;&#32;&#100;&#x6f;&#116;&#32;&#x63;&#x6f;&#x6d;</noscript></li>
<li><script type="text/javascript">
<!--
h='&#112;&#104;&#x61;&#114;&#x6d;&#x61;&#122;&#x69;&#x65;&#46;&#x75;&#110;&#x69;&#x2d;&#102;&#114;&#x65;&#x69;&#98;&#x75;&#114;&#x67;&#46;&#100;&#x65;';a='&#64;';n='&#x6b;&#x65;&#114;&#x73;&#116;&#x65;&#110;&#46;&#100;&#x6f;&#x65;&#114;&#x69;&#110;&#x67;';e=n+a+h;
document.write('<a h'+'ref'+'="ma'+'ilto'+':'+e+'">'+e+'<\/'+'a'+'>');
// -->
</script><noscript>&#x6b;&#x65;&#114;&#x73;&#116;&#x65;&#110;&#46;&#100;&#x6f;&#x65;&#114;&#x69;&#110;&#x67;&#32;&#x61;&#116;&#32;&#112;&#104;&#x61;&#114;&#x6d;&#x61;&#122;&#x69;&#x65;&#32;&#100;&#x6f;&#116;&#32;&#x75;&#110;&#x69;&#x2d;&#102;&#114;&#x65;&#x69;&#98;&#x75;&#114;&#x67;&#32;&#100;&#x6f;&#116;&#32;&#100;&#x65;</noscript></li>
</ul>
</blockquote></li>
<li><p>If you are interested in related projects, visit our working group’s homepage:</p>
<blockquote>
<ul>
<li><a href="http://www.pharmaceutical-bioinformatics.de">http://www.pharmaceutical-bioinformatics.de</a></li>
</ul>
</blockquote></li>
<li>This project is published with an ISC license given in “LICENSE.md”.</li>
</ul>
</body>
</html>
